test_plans/exceptions.asciidoc

```diff
@@ -0,0 +1,243 @@
+:sectnums:
+:xrefstyle: short
+
+= Test plan for exceptions
+
+This is a test plan for the APIs described in SYCL 2020 section 4.13.2.
+"Exception class interface"
+
+== Testing scope
+
+=== Device coverage
+
+All of the tests described below are performed only on the default device that
+is selected on the CTS command line.
+
+
+== Tests
+
+Every method should be checked for accordance to spec:
+
+* Method existance
+* Const correctness
+* Type of result
+* Result for noexcept() operator
+
+=== sycl namespace functions
+
+==== make_error_code
+
+For every errc value check that:
+
+* Result of the call should be equal to `std::error_code(errc)`
+* `error_code::value` is equal to errc value passed
+* `error_code::category` is equal to sycl_category()
+* `error_code::default_error_condition()` is equal to
+  `std::error_condition(errc, sycl::sycl_category())`
+
+==== sycl_category()
+
+For each of the following use-cases:
+
+* using reference or copy of `error_category` retrieved at the function scope
+* using static variable in the test namespace that is constructed from sycl_category()
+  and then copied
+
+provide checks under a separate test name:
+
+* `error_category::name()` should return "sycl"
+* For `sycl::errc::success` and any other one `errc` value `error_category::default_error_condition(errc)`
+  is equal to `std::err_condition(errc, sycl::sycl_category())`
+* For FULL conformance - same as bullet above but for all `errc` values
+
+==== error_category_for()
+
+* Check that function template exists.
+
+If `SYCL_BACKEND_OPENCL` is defined run same checks as for sycl_category, but:
+
+* `error_category::name()` should return "opencl"
+* for value = static_cast<sycl::errc_for<sycl::backend::opencl>>(0)
+  `error_category_for<>::default_error_condition(value)` should be equal to
+  `std::error_condition(value, sycl::error_category_for<sycl::backend::opencl>())`
+
+=== sycl::errc
+
+For all values from spec:
+
+* `runtime`
+* `kernel`
+* `accessor`
+* `nd_range`
+* `event`
+* `kernel_argument`
+* `build`
+* `invalid`
+* `memory_allocation`
+* `platform`
+* `profiling`
+* `feature_not_supported`
+* `kernel_not_supported`
+* `backend_mismatch`
+
+check that:
+
+* that they exsist
+* that all values are different
+
+Special cases:
+
+* Check that `std::is_error_code_enum` is `true` for `sycl::errc`
+* Check that `std::is_error_condition_enum` should be `false` for `sycl::err`
+
+=== sycl::errc_for
+
+* Check that enum class template exists
+
+If SYCL_BACKEND_OPENCL is defined check that
+
+* enum class with `b == sycl::backend::opencl` exists
+* there is value 0
```

Comment:
looks like it's not possible to verify the C++ enum contains specific integral value, but still we can verify that `errc_for<sycl::backend::opencl>` is a scoped enum type
if `errc_for` is a scoped enumeration type we are safe to use **0** value for scoped enums